src/lib/protocols/ieee-c37118.c

```diff
@@ -0,0 +1,93 @@
+/*
+ * ieee-c37118.c
+ *
+ * IEEE C37.118 Synchrophasor Protocol
+ * 
+ * Copyright (C) 2023 - ntop.org
+ * Copyright (C) 2023 - V.G <jacendi@protonmail.com>
+ *
+ * This file is part of nDPI, an open source deep packet inspection
+ * library based on the OpenDPI and PACE technology by ipoque GmbH
+ *
+ * nDPI is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * nDPI is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with nDPI.  If not, see <http://www.gnu.org/licenses/>.
+ * 
+ */
+
+#include "ndpi_protocol_ids.h"
+
+#define NDPI_CURRENT_PROTO NDPI_PROTOCOL_IEEE_C37118
+
+#include "ndpi_api.h"
+#include "ndpi_private.h"
+
+static u_int16_t crc16(const u_int8_t* data, size_t n_bytes)
+{
+  u_int16_t crc = 0xFFFF;
+  
+  while (n_bytes--) {
+    crc ^= *data++ << 8;
+    for (u_int8_t i = 0; i < 8; i++) {
+      crc = crc & 0x8000 ? (crc << 1) ^ 0x1021 : crc << 1;
+    }
+  }
+  
+  return crc & 0xFFFF;
+}
```

Comment:
Not too optimal, but it works. I guess this could be replaced by the optimal lookup table CRC16 algorithm if it's ever added to nDPI.

Reply:
> Could you open an issue about that, so we don't forget, please?

Sure. But tbh, I currently only know of one protocol that also uses CRC16 CCITT and that's Saia-Burgess Electronics Ether-S-Bus.